### PR TITLE
Authorization for expand-spec, InstanceTypes and Preferences 

### PR DESCRIPTION
**Pull request by 0xFelix**
**What this PR does / why we need it**:

This is adding authorization for the non-namespaced `expand-spec` subresource endpoint as well as for InstanceTypes and Preferences themselves.

Based on https://github.com/kubevirt/kubevirt/pull/8399

/area instancetype

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Added authorization to the namespaced and non-namespaced expand-spec subresource endpoints
Added authorization to InstanceTypes and Preferences referenced in expand-spec subresource requests
```

### CODE REVIEW

**Comment by 0xFelix** [comment]
@akrejcir You're right, but now it's in a state where I would like to see what the CI says.

**Comment by akrejcir** [comment]
> @akrejcir You're right, but now it's in a state where I would like to see what the CI says.

You can still trigger test lanes manually, if you don't need to run all of them.

File: pkg/virt-api/rest/authorizer.go
Comment (akrejcir) [inline pkg/virt-api/rest/authorizer.go:31]: Can you move all refactoring to a separate commit? I think it would be easier to see new functionality.

File: pkg/virt-api/rest/authorizer.go
Comment (akrejcir) [inline pkg/virt-api/rest/authorizer.go:223]: Can you add a comment here explaining that `expand-spec` is treated as a resource name by k8s, even though there is no CRD that defines it?

File: pkg/virt-api/rest/authorizer_test.go
Comment (akrejcir) [inline pkg/virt-api/rest/authorizer_test.go:29]: Similarly, can you move the refactoring to a separate commit?

File: pkg/virt-api/rest/authorizer.go
Comment (akrejcir) [inline pkg/virt-api/rest/authorizer.go:53]: Can you squash this commit with the one where these changes are useful? Then it will be easier to review.
This commit by itself does not seem useful.

File: pkg/virt-api/rest/authorizer.go
Comment (akrejcir) [inline pkg/virt-api/rest/authorizer.go:152]: I think this method can be removed. It only calls the client inside `authorizor`, which is passed to method `NewAuthorizorFromClient()`. A potential caller of this method can instead call the client directly.

**Review by akrejcir — CHANGES_REQUESTED** [review]

File: pkg/virt-api/rest/authorizer.go
Comment (0xFelix) [reply pkg/virt-api/rest/authorizer.go:223]: Good point!

File: pkg/virt-api/rest/expand.go
Comment (0xFelix) [reply pkg/virt-api/rest/expand.go:90]: Done

File: pkg/virt-api/rest/authorizer.go
Comment (0xFelix) [reply pkg/virt-api/rest/authorizer.go:53]: Done

File: pkg/virt-api/rest/authorizer.go
Comment (0xFelix) [reply pkg/virt-api/rest/authorizer.go:31]: Done

File: pkg/instancetype/instancetype.go
Comment (lyarwood) [reply pkg/instancetype/instancetype.go:32]: I think that's a little over overzealous, it's functionality specific to instancetypes and preferences so this is a perfectly reasonible home for the code IMHO. Especially given that we already have instancetypemethods available in virt-api/rest.

File: pkg/instancetype/instancetype.go
Comment (lyarwood) [inline pkg/instancetype/instancetype.go:413]: Given we don't need to actually look anything up here given @akrejcir other comments could you also rename this to `CreatePreferenceResourceAttributes` ?

File: pkg/instancetype/instancetype.go
Comment (lyarwood) [inline pkg/instancetype/instancetype.go:95]: This should really live somewhere more generic but that's something for a followup.

File: pkg/virt-api/rest/authorizer_test.go
Comment (0xFelix) [reply pkg/virt-api/rest/authorizer_test.go:29]: Done

File: pkg/instancetype/instancetype.go
Comment (akrejcir) [reply pkg/instancetype/instancetype.go:32]: Ok I see, it can stay here. But can you move it outside of the `Methods` interface and make it a free function?

Comment (lyarwood) [reply pkg/instancetype/instancetype.go:32]: Yeah good point, given that we don't need to look anything up anymore that's valid.

Comment (0xFelix) [reply pkg/instancetype/instancetype.go:32]: Done

File: pkg/instancetype/instancetype.go
Comment (0xFelix) [reply pkg/instancetype/instancetype.go:420]: Good call, I changed it.

File: pkg/virt-api/rest/expand.go
Comment (0xFelix) [reply pkg/virt-api/rest/expand.go:83]: Done

File: pkg/instancetype/instancetype.go
Comment (0xFelix) [reply pkg/instancetype/instancetype.go:95]: I dropped it for now as it is no longer needed inside this series.

File: pkg/instancetype/instancetype.go
Comment (0xFelix) [reply pkg/instancetype/instancetype.go:413]: Done

**Comment by lyarwood** [comment]
/lgtm
Thanks @0xFelix!

File: pkg/virt-api/rest/expand.go
Comment (akrejcir) [inline pkg/virt-api/rest/expand.go:96]: Do you think it makes sense to move this line before calling `FindInstancetypeSpec()`?
It seems unusual to me that permission to access instancetype is checked after accessing the instancetype.

File: pkg/virt-api/rest/expand.go
Comment (akrejcir) [inline pkg/virt-api/rest/expand.go:129]: See my above comment.

File: pkg/virt-api/rest/expand.go
Comment (0xFelix) [reply pkg/virt-api/rest/expand.go:96]: The order was intentionally swapped to prevent authorization checks on non-existent objects. WDYT?

Comment (akrejcir) [reply pkg/virt-api/rest/expand.go:96]: I'm not sure. I will leave the decision to the approver.

Comment (lyarwood) [reply pkg/virt-api/rest/expand.go:96]: /lgtm cancel

Thinking about this some more wouldn't this ordering potentially leak the fact that a resource existed or not without the user necessarily having the correct privileges to see it?

Appreciate this is just instancetypes and preferences so leaking that to the caller might not be the end of the world but reordering to ensure they have access before checking for the resource might be cleaner.

Comment (0xFelix) [reply pkg/virt-api/rest/expand.go:90]: Fair enough, you are right, it could leak the fact a resource doesn't exist. I swapped the order again, so authorization checks come first.

**Review by lyarwood — COMMENTED** [review]
/lgtm

**Comment by lyarwood** [comment]
> This looks like way more than I'd expect to be needed. why isn't the authorization filter catching this?
> 
> https://github.com/kubevirt/kubevirt/blob/52a3f52ee14c14a6cc58f5bb16ebb8784d686c9a/pkg/virt-api/api.go#L684

Yeah the additional complexity here is around ensuring a caller to the `expand-spec` API is authorised to interact with instancetypes and preferences referenced within the `VirtualMachine` they are providing within their request. 

I appreciate it's slightly over the top checking this at the moment given we don't support cross namespace references for either instancetypes or preferences from the VirtualMachine but I'd like to see this covered before we potentially do in the future.

**Comment by davidvossel** [comment]
> Yeah the additional complexity here is around ensuring a caller to the expand-spec API is authorised to interact with instancetypes and preferences referenced within the VirtualMachine they are providing within their request.

oh, so the issue is that the "expand-spec" endpoint isn't namespaced which allows people to use this cluster scoped api endpoint to get instancetype/preferences from any namespace.

What about making "expand-spec" a namespace scoped subresource endpoint?

**Comment by lyarwood** [comment]
> > Yeah the additional complexity here is around ensuring a caller to the expand-spec API is authorised to interact with instancetypes and preferences referenced within the VirtualMachine they are providing within their request.
> 
> oh, so the issue is that the "expand-spec" endpoint isn't namespaced which allows people to use this cluster scoped api endpoint to get instancetype/preferences from any namespace.
> 
> What about making "expand-spec" a namespace scoped subresource endpoint?

That would still only authorise the user in terms of the provided `VirtualMachine` and assumes that the user has access to any referenced instancetype or preference resources in provided namespace right? I appreciate that this is inline with our current behaviour when creating a `VirtualMachine` but I believe @0xFelix wanted to look into also authorising these requests more rigorously once this PR had landed.

FWIW I've also had requests to expand the `Instancetype,Preference{Matchers}` to allow cross namespace references in the future (maybe even `v1beta1`) so the authorisation functionality introduced by this PR would definitely be required to support that use case.

**Comment by 0xFelix** [comment]
Making the `expand-spec` subresource namespace scoped would indeed suffice to limit users to expand instancetypes and namespaces in the current namespace at the moment. Access to cluster wide instancetypes and namespaces could not be restricted. 

In regard to @lyarwood plans to allow cross namespacing in the future making `expand-spec` namespace scoped wouldn't suffice anymore. Also what would differentiate a namespace scoped `expand-spec` from a cluster scoped `expand-spec` when cross namespacing is possible?

**Comment by davidvossel** [comment]
here are some more thoughts

> FWIW I've also had requests to expand the Instancetype,Preference{Matchers} to allow cross namespace references in the future (maybe even v1beta1)

@lyarwood I think it is doubtful we'll do cross namespace references. This is the whole reason we made global instancetypes/preferences, so we don't have to try and figure out cross namespace permissions.

We made cross namespace PVC cloning simply because we had to. There's no such thing as a global PVC, and PVCs were the thing we decided to use to represent disk images, so cross namespace was practically our only option if we wanted to make "golden images" available across all namespaces. It was very complicated to get right ( hopefully we got it right :D )

> Making the expand-spec subresource namespace scoped would indeed suffice to limit users to expand instancetypes and namespaces in the current namespace at the moment. Access to cluster wide instancetypes and namespaces could not be restricted.

If we transition to a namespaced expand-spec, I think that solves a few issues.
- One, we now naturally have RBAC per namespace ensuring that users can only gain info about instancetypes/preferences from the namespace they have access to.

- Two, it solves so weird edge cases. For example, what if someone references a namespaced instancetype but doesn't set the namespace in their VM spec. That VM would technically work if Created because the namespace is inferred at creation to the default namespace for the user. However with a global expand-spec endpoint, we can't infer that... so it's weird. With a namespaced expand-spec, we can infer the namespace accurately.

**Comment by lyarwood** [comment]
> here are some more thoughts
> 
> > FWIW I've also had requests to expand the Instancetype,Preference{Matchers} to allow cross namespace references in the future (maybe even v1beta1)
> 
> @lyarwood I think it is doubtful we'll do cross namespace references. This is the whole reason we made global instancetypes/preferences, so we don't have to try and figure out cross namespace permissions.
> 
> We made cross namespace PVC cloning simply because we had to. There's no such thing as a global PVC, and PVCs were the thing we decided to use to represent disk images, so cross namespace was practically our only option if we wanted to make "golden images" available across all namespaces. It was very complicated to get right ( hopefully we got it right :D )

ACK understood, FWIW the request for cross namespace references was born from the current common-templates where we don't have global resources and instead had a shared namespace that the SSP operator would essentially make immutable. As you say we can do this with out global resources, something I had attempted to push when this initially came up and will do so again.

> > Making the expand-spec subresource namespace scoped would indeed suffice to limit users to expand instancetypes and namespaces in the current namespace at the moment. Access to cluster wide instancetypes and namespaces could not be restricted.
> 
> If we transition to a namespaced expand-spec, I think that solves a few issues.
> 
> * One, we now naturally have RBAC per namespace ensuring that users can only gain info about instancetypes/preferences from the namespace they have access to.

ACK true.

> * Two, it solves so weird edge cases. For example, what if someone references a namespaced instancetype but doesn't set the namespace in their VM spec. That VM would technically work if Created because the namespace is inferred at creation to the default namespace for the user. However with a global expand-spec endpoint, we can't infer that... so it's weird. With a namespaced expand-spec, we can infer the namespace accurately.

Yup it definitely simplifies the entire approach. @0xFelix, @akrejcir  you've both but a huge amount of work into this PR so I'd appreciate it if you could comment on this and agree before we change direction here.

**Comment by 0xFelix** [comment]
This PR was a good learning experience and it might come in handy in the future. If we move to a namespaced `expand-spec` subresource endpoint and drop the requirement for cross namespacing, weirdnesses like David pointed out can be avoided and the impact on the code will also be much lower. If we are sure we need no cross namespacing IMO we can go for it.

**Comment by 0xFelix** [comment]
Closing in favor of https://github.com/kubevirt/kubevirt/pull/8720

**Comment by akrejcir** [comment]
I agree, a namespaced resource makes sense. 

When I wrote the initial PR, I incorrectly assumed that k8s API extension can define subresources only on existing CRDs, or cluster wide endpoints:
```
/apis/subresources.kubevirt.io/v1/namespaces/${NAMESPACE}/virtualmachine/${VM_NAME}/expand-spec
/apis/subresources.kubevirt.io/v1/expand-spec
```

If it is possible to define:
```
/apis/subresources.kubevirt.io/v1/namespaces/${NAMESPACE}/expand-vm-spec
```
then we can use it.

**Pull request closed without merging**
